worker/src/monitor/metrics_collector.rs

```diff
@@ -0,0 +1,67 @@
+use std::{thread::sleep, time::Duration};
+
+use proto::common::node::Metrics;
+use sysinfo::System;
+
+const CPU_DELAY_IN_MILLIS: u64 = 500;
+
+/// Responsable for collecting metrics about the System, such as
+/// CPU Percentage Usage, Free Memory Space.
+pub struct MetricsCollector {
+    system: System,
+}
+impl MetricsCollector {
+    /// Instantiates a new `MetricsCollector`.
+    pub fn new() -> Self {
+        Self {
+            system: System::new_all(),
+        }
+    }
+
+    /// Get the `cpu_usages` and `free_memory` metrics from the current system.
+    ///
+    /// When calling this method, will sleep the thread for
+    /// `CPU_DELAY_IN_MILLIS`.
+    pub fn get_metrics(&mut self) -> Metrics {
+        Metrics {
+            cpu_usage: self.get_cpu_usage(),
+            mem_total_mib: self.get_total_memory(),
+            mem_used_mib: self.get_used_memory(),
+        }
+    }
+
+    fn get_cpu_usage(&mut self) -> f64 {
+        self.system.refresh_cpu();
+        sleep(Duration::from_millis(CPU_DELAY_IN_MILLIS));
+        self.system.refresh_cpu();
+
+        let cpus_usages: Vec<f64> = self
+            .system
+            .cpus()
+            .iter()
+            .map(|cpu| f64::from(cpu.cpu_usage()))
+            .collect();
```

Comment:
You can call `sum` directly here to avoid the vector allocation :)

Reply:
True! I need the length of the `Vec` before reducing it to a single value though. So I came up with this:: 
```rs
        let amount_cpus: usize = self
            .system
            .cpus()
            .len();

        let sum_cpus_usages: f64 = self
            .system
            .cpus()
            .iter()
            .map(|cpu| f64::from(cpu.cpu_usage()))
            .sum(); 

        sum_cpus_usages / amount_cpus as f64
    }
 ```